static/js/vdom.js

```diff
@@ -0,0 +1,132 @@
+exports.eq_array = (a, b, eq) => {
+    if (a === b) {
+        // either both are undefined, or they
+        // are referentially equal
+        return true;
+    }
+
+    if (a === undefined || b === undefined) {
+        return false;
+    }
+
+    if (a.length !== b.length) {
+        return false;
+    }
+
+    return _.all(a, (item, i) => {
+        return eq(item, b[i]);
+    });
+};
+
+exports.ul = (opts) => {
+    return {
+        tag_name: 'ul',
+        opts: opts,
+    };
+};
+
+exports.render = (tag) => {
+    const opts = tag.opts;
+    const tag_name = tag.tag_name;
+    const attr_str = _.map(opts.attrs, (attr) => {
+        return ' ' + attr[0] + '="' + attr[1] + '"';
+    }).join('');
+
+    const start_tag = '<' + tag_name + attr_str + '>';
+    const end_tag = '</' + tag_name + '>';
+
+    if (opts.keyed_nodes === undefined) {
+        blueslip.error("We need keyed_nodes to render innards.");
+        return;
+    }
+
+    const innards = _.map(opts.keyed_nodes, (node) => {
+        return node.render();
+    }).join('\n');
+    return start_tag + '\n' + innards + '\n' + end_tag;
+};
+
+exports.update = (container, new_dom, old_dom) => {
```

Comment:
Let's add a block comment here explaining what this supports and what cases it optimizes well.

Reply:
done